File: src/lib/api/tags.ts
import { invoke } from "@tauri-apps/api/core";
import type {
  TagItem,
  TagCreateInput,
  TagUpdateInput,
  TagReorderInput,
} from "../../types/tag";

const STORAGE_KEY = "bugoo:tags";

function getLocalTags(): TagItem[] {
  const saved = localStorage.getItem(STORAGE_KEY);
  return saved ? JSON.parse(saved) : [];
}

function setLocalTags(tags: TagItem[]): void {
  localStorage.setItem(STORAGE_KEY, JSON.stringify(tags));
}

function generateId(): string {
  return `tag_${Date.now()}_${Math.random().toString(36).slice(2, 8)}`;
}

/**
 * 获取所有标签（按 sort_order 升序）
 * 后端未就绪时自动降级到 localStorage
 */
export async function getTags(): Promise<TagItem[]> {
  try {
    return await invoke<TagItem[]>("get_tags");
  } catch (e) {
    console.error("[tags] invoke('get_tags') failed, falling back to localStorage:", e);
    return getLocalTags().sort((a, b) => a.sort_order - b.sort_order);
  }
}

export async function createTag(input: TagCreateInput): Promise<TagItem> {
  try {
    return await invoke<TagItem>("create_tag", { input });
  } catch {
    const newTag: TagItem = {
      id: generateId(),
      name: input.name,
      color: input.color,
      sort_order: input.sort_order ?? Date.now(),
      created_at: Date.now(),
      updated_at: Date.now(),
    };
    const tags = getLocalTags();
    tags.push(newTag);
    setLocalTags(tags);
    return newTag;
  }
}

export async function updateTag(
  id: string,
  input: TagUpdateInput,
): Promise<TagItem> {
  try {
    return await invoke<TagItem>("update_tag", { id, input });
  } catch {
    const tags = getLocalTags();
    const index = tags.findIndex((t) => t.id === id);
    if (index === -1) throw new Error(`Tag ${id} not found`);
    tags[index] = { ...tags[index], ...input, updated_at: Date.now() };
    setLocalTags(tags);
    return tags[index];
  }
}

export async function deleteTag(id: string): Promise<void> {
  try {
    await invoke<void>("delete_tag", { id });
  } catch {
    const tags = getLocalTags();
    const filtered = tags.filter((t) => t.id !== id);
    setLocalTags(filtered);
  }
}

export async function reorderTags(input: TagReorderInput): Promise<TagItem[]> {
  try {
    return await invoke<TagItem[]>("reorder_tags", { input });
  } catch {
    const tags = getLocalTags();
    const tagMap = new Map(tags.map((t) => [t.id, t]));
    const reordered = input.tag_ids.map((id, index) => {
      const tag = tagMap.get(id);
      if (!tag) throw new Error(`Tag ${id} not found`);
      return { ...tag, sort_order: index, updated_at: Date.now() };
    });
    setLocalTags(reordered);
    return reordered;
  }
}

File: src-tauri/src/commands/tags.rs
use crate::commands::AppState;
use crate::domain::models::tag::{Tag, TagCreateInput, TagUpdateInput, TagReorderInput};
use crate::db::tags as db_tags;

/// 获取所有标签
#[tauri::command]
pub fn get_tags(state: tauri::State<AppState>) -> Result<Vec<Tag>, String> {
    let conn = state.db.connection();
    db_tags::get_all_tags(&conn).map_err(|e| e.to_string())
}

/// 创建新标签
#[tauri::command]
pub fn create_tag(
    state: tauri::State<AppState>,
    input: TagCreateInput,
) -> Result<Tag, String> {
    let conn = state.db.connection();

    // 检查名称唯一性
    if let Some(existing) = db_tags::find_tag_by_name(&conn, &input.name).map_err(|e| e.to_string())? {
        return Err(format!("Tag '{}' already exists", existing.name));
    }

    db_tags::create_tag(&conn, input).map_err(|e| e.to_string())
}

/// 更新标签
#[tauri::command]
pub fn update_tag(
    state: tauri::State<AppState>,
    id: String,
    input: TagUpdateInput,
) -> Result<Tag, String> {
    let conn = state.db.connection();

    // 如果更新名称，检查唯一性
    if let Some(new_name) = &input.name {
        if let Some(existing) = db_tags::find_tag_by_name(&conn, new_name).map_err(|e| e.to_string())? {
            if existing.id != id {
                return Err(format!("Tag '{}' already exists", existing.name));
            }
        }
    }

    db_tags::update_tag(&conn, &id, input).map_err(|e| e.to_string())
}

/// 删除标签
#[tauri::command]
pub fn delete_tag(
    state: tauri::State<AppState>,
    id: String,
) -> Result<(), String> {
    let conn = state.db.connection();
    db_tags::delete_tag(&conn, &id).map_err(|e| e.to_string())?;
    Ok(())
}

/// 重排标签顺序
#[tauri::command]
pub fn reorder_tags(
    state: tauri::State<AppState>,
    input: TagReorderInput,
) -> Result<Vec<Tag>, String> {
    let conn = state.db.connection();
    db_tags::reorder_tags(&conn, input.tag_ids).map_err(|e| e.to_string())
}